scripts/gpu_support/nvidia/install_cuda_and_libraries.sh

```diff
@@ -0,0 +1,205 @@
+    echo "  -t, --temp-dir /path/to/tmpdir   Specify a location to use for temporary"
+    echo "                                   storage during the installation of CUDA"
+    echo "                                   and/or other libraries (must have"
+    echo "                                   several GB available; depends on the number of installations)"
+}
+
+# Initialize variables
+eula_accepted=0
+EASYSTACKFILE=
+TEMP_DIR=
+
+# Parse command-line options
+while [[ $# -gt 0 ]]; do
+    case "$1" in
+        --help)
+            show_help
+            exit 0
+            ;;
+        --accept-cuda-eula)
+            eula_accepted=1
+            shift 1
+            ;;
+        -e|--easystack)
+            if [ -n "$2" ]; then
+                EASYSTACKFILE="$2"
+                shift 2
+            else
+                echo "Error: Argument required for $1"
+                show_help
+                exit 1
+            fi
+            ;;
+        -t|--temp-dir)
+            if [ -n "$2" ]; then
+                TEMP_DIR="$2"
+                shift 2
+            else
+                echo "Error: Argument required for $1"
+                show_help
+                exit 1
+            fi
+            ;;
+        *)
+            show_help
+            fatal_error "Error: Unknown option: $1"
+            ;;
+    esac
+done
+
+if [[ -z "${EASYSTACKFILE}" ]]; then
+    fatal_error "Need the name/path to an easystack file. See command line options\n"
+fi
+
+# Make sure EESSI is initialised
+check_eessi_initialised
+
+# As an installation location just use $EESSI_SOFTWARE_PATH but replacing `versions` with `host_injections`
+# (CUDA is a binary installation so no need to worry too much about the EasyBuild setup)
+export EESSI_SITE_INSTALL=${EESSI_SOFTWARE_PATH/versions/host_injections}
+
+# we need a directory we can use for temporary storage
+if [[ -z "${TEMP_DIR}" ]]; then
+    tmpdir=$(mktemp -d)
+else
+    mkdir -p ${TEMP_DIR}
+    tmpdir=$(mktemp -d --tmpdir=${TEMP_DIR} cuda_n_co.XXX)
+    if [[ ! -d "$tmpdir" ]] ; then
+        fatal_error "Could not create directory ${tmpdir}"
+    fi
+fi
+echo "Created temporary directory '${tmpdir}'"
+
+# workaround for EasyBuild not being found when loading "extend" module
+module load EasyBuild/4.9.4
+
+# load EESSI-extend/2023.06-easybuild module && verify that it is loaded
+EESSI_EXTEND_MODULE="EESSI-extend/2023.06-easybuild"
+module load ${EESSI_EXTEND_MODULE}
+ret=$?
+if [ "${ret}" -ne 0 ]; then
+    fatal_error "An error occured while trying to load ${EESSI_EXTEND_MODULE}\n"
+fi
+
+# do a 'eb --dry-run-short' with the EASYSTACKFILE and determine list of packages
+# to be installed
+echo ">> Determining if packages specified in ${EASYSTACKFILE} are missing under ${EESSI_SITE_INSTALL}"
+eb_dry_run_short_out=${tmpdir}/eb_dry_run_short.out
+eb --dry-run-short --rebuild --easystack ${EASYSTACKFILE} 2>&1 | tee ${eb_dry_run_short_out}
```

Comment:
What's the output of this? I'm surprised that this would work. The CUDA install (for example) under `${EESSI_SITE_INSTALL}` has no module file, so for EasyBuild I  expect that it does not exist.

What I thought we would have to do here is to do a `--module-only` build first (accepting that something may fail), and then follow it with a normal build for the gaps.

Reply:
Output of the command seems to be or include the following
```
>> Determining if packages specified in /cvmfs/software.eessi.io/versions/2023.06/scripts/gpu_support/nvidia/eessi-2023.06-cuda-and-libraries.yml are missing under /cvmfs/software.eessi.io/host_injections/2023.06/software/linux/x86_64/amd/zen3
== Temporary log file in case of crash /tmp/eb-5tv910kz/easybuild-gycubict.log
== found valid index for /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/EasyBuild/4.9.4/easybuild/easyconfigs, so using it...
== Running parse hook for CUDA-12.1.1.eb...
Dry run: printing build status of easyconfigs and dependencies
CFGS=/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/EasyBuild/4.9.4/easybuild/easyconfigs/c/CUDA
 * [R] $CFGS/CUDA-12.1.1.eb (module: CUDA/12.1.1)
== found valid index for /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/EasyBuild/4.9.4/easybuild/easyconfigs, so using it...
== Running parse hook for cuDNN-8.9.2.26-CUDA-12.1.1.eb...
== found valid index for /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/EasyBuild/4.9.4/easybuild/easyconfigs, so using it...
== Running parse hook for CUDA-12.1.1.eb...
Dry run: printing build status of easyconfigs and dependencies
CFGS=/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/EasyBuild/4.9.4/easybuild/easyconfigs/c
 * [x] $CFGS/CUDA/CUDA-12.1.1.eb (module: CUDA/12.1.1)
 * [ ] $CFGS/cuDNN/cuDNN-8.9.2.26-CUDA-12.1.1.eb (module: cuDNN/8.9.2.26-CUDA-12.1.1)
== Temporary log file(s) /tmp/eb-5tv910kz/easybuild-gycubict.log* have been removed.
== Temporary directory /tmp/eb-5tv910kz has been removed.
```

Reply:
Ah, this is only checking what is available in the MODULEPATH, which CUDA is because it is available in the shipped EESSI. This is not actually checking what is available under `host_injections`.

There is a design choice to be made here. Before I didn't ship the final module files for CUDA, because I only cared about the fixing the symlinks. Now, there is an argument to be made that we _should_ ship the module files for the installations under `host_injections` as they are automatically picked up by our initialisation scripts:
- In any scenario we need to start with an empty `MODULEPATH` when performing this check
- If we don't ship the module files, we need to regenerate them (after clearing out `MODULEPATH`) before we can reliably check whether the easyconfigs in the easystack file are present or not under `host_injections`
- We could just ship the modules, but there is a chance this will be confusing for people (and any software they build themselves will link against the `host_injections` runtime _not_ the EESSI runtime)
- We could put them in a different location under `host_injections` that is not exposed to end users, that way EasyBuild can still identify if they have been installed or not without additional tricks

Reply:
Personal choice would be option 3

Reply:
3 with a hidden `.modules` subdirectory maybe?

This seems orthogonal to this PR though? If so, let's open a dedicated issue on this

Reply:
Then that PR would need to be merged first, as we're relying it to resolve the symlinks

Reply:
As it stands, running this script will not correctly recognise what has and has not been installed under `host_injections`. I think this is a nice approach though, I like the easystack file to keep track of things.

Reply:
We'll, it's true that this is really an optimisation, you can of course rebuild everything every time you run the script (which is what I understand is happening now). We can improve that later.

Reply:
implemented option 3 in https://github.com/EESSI/software-layer/pull/772/commits/e3101b0c82d554823b33d9cab2abee074e330222

Reply:
It almost works as wanted...

Job in https://github.com/EESSI/software-layer/pull/772#issuecomment-2394695882 installed CUDA and cuDNN and the module files under `.../.modules/all`

The check in job https://github.com/EESSI/software-layer/pull/772#issuecomment-2394728700 found the installed modules.

However, both CUDA and cuDNN where then reinstalled because we run EasyBuild with `--rebuild`. If we want to avoid the reinstallation, we should remove `--rebuild` (at least for the actual installation run; I guess it can also be removed from the run that determines how many packages would be installed).